File: app/src/main/java/net/ducksmanager/whattheduck/AddIssue.java
package net.ducksmanager.whattheduck;

import android.app.DatePickerDialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;

import net.ducksmanager.apigateway.DmServer;
import net.ducksmanager.persistence.models.composite.InducksIssueWithUserIssueDetails;
import net.ducksmanager.persistence.models.composite.IssueListToUpdate;
import net.ducksmanager.persistence.models.dm.Purchase;

import java.lang.ref.WeakReference;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import retrofit2.Response;

import static net.ducksmanager.whattheduck.WhatTheDuck.appDB;
import static net.ducksmanager.whattheduck.WhatTheDuck.info;
import static net.ducksmanager.whattheduck.WhatTheDuck.selectedIssue;
import static net.ducksmanager.whattheduck.WhatTheDuck.selectedPublication;

public class AddIssue extends AppCompatActivity implements View.OnClickListener {

    private List<Purchase> purchases;

    private static final Calendar myCalendar = Calendar.getInstance();
    private static final SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.addissue);
        downloadPurchaseList();
    }

    private void downloadPurchaseList() {
        DmServer.api.getUserPurchases().enqueue(new DmServer.Callback<List<Purchase>>("getpurchases", this, true) {
            @Override
            public void onSuccessfulResponse(Response<List<Purchase>> response) {
                appDB.purchaseDao().deleteAll();
                appDB.purchaseDao().insertList(response.body());
                setData();
            }
        });
    }

    private void setData() {
        appDB.purchaseDao().findAll().observe(this, purchases -> {
            this.purchases = new ArrayList<>();
            this.purchases.add(new PurchaseAdapter.NoPurchase());
            this.purchases.addAll(purchases);
            this.show();
        });
    }

    private void show() {
        findViewById(R.id.noCondition).setOnClickListener(this);
        findViewById(R.id.badCondition).setOnClickListener(this);
        findViewById(R.id.notSoGoodCondition).setOnClickListener(this);
        findViewById(R.id.goodCondition).setOnClickListener(this);

        showPurchases();
        findViewById(R.id.noCondition).performClick();
        PurchaseAdapter.selectedItem = purchases.get(0);

        findViewById(R.id.addissue_ok).setOnClickListener(view -> {
            String dmCondition;
            RadioGroup r = findViewById(R.id.condition);
            switch(r.getCheckedRadioButtonId()) {
                case R.id.badCondition: dmCondition = InducksIssueWithUserIssueDetails.BAD_CONDITION; break;
                case R.id.notSoGoodCondition: dmCondition = InducksIssueWithUserIssueDetails.NOTSOGOOD_CONDITION; break;
                case R.id.goodCondition: dmCondition = InducksIssueWithUserIssueDetails.GOOD_CONDITION; break;
                default: dmCondition = InducksIssueWithUserIssueDetails.NO_CONDITION; break;
            }

            IssueListToUpdate issueListToUpdate = new IssueListToUpdate(
                selectedPublication,
                Collections.singletonList(selectedIssue),
                dmCondition,
                PurchaseAdapter.selectedItem instanceof PurchaseAdapter.NoPurchase
                    ? null
                    : PurchaseAdapter.selectedItem.getId()
            );

            DmServer.api.createUserIssues(issueListToUpdate).enqueue(new DmServer.Callback<Object>("addissue", AddIssue.this, true) {
                @Override
                public void onSuccessfulResponse(Response<Object> response) {
                    finish();
                    info(new WeakReference<>(AddIssue.this), R.string.confirmation_message__issue_inserted, Toast.LENGTH_SHORT);
                    startActivity(new Intent(AddIssue.this, Login.class));
                }
            });
        });

        findViewById(R.id.addissue_cancel).setOnClickListener(view ->
            finish()
        );

        findViewById(R.id.addpurchase).setOnClickListener(view -> {
            toggleAddPurchaseButton(false);
            showNewPurchase();
        });

        ((TextView)findViewById(R.id.addIssueTitle)).setText(getString(R.string.insert_issue__confirm, selectedIssue));
    }

    private void toggleAddPurchaseButton(Boolean toggle) {
        findViewById(R.id.addpurchase).setVisibility(toggle ? View.VISIBLE : View.GONE);
    }
    
    private void showNewPurchase() {
        View newPurchaseSection = findViewById(R.id.newpurchase);
        EditText purchaseDateNew = findViewById(R.id.purchasedatenew);
        EditText purchaseTitleNew = findViewById(R.id.purchasetitlenew);
        Button purchaseCreate = findViewById(R.id.createpurchase);
        Button purchaseCreateCancel = findViewById(R.id.createpurchasecancel);

        newPurchaseSection.setVisibility(View.VISIBLE);
        purchaseDateNew.requestFocus();
        purchaseDateNew.setText(dateFormat.format(new Date()));
        purchaseDateNew.setKeyListener(null);

        final DatePickerDialog.OnDateSetListener date = (datePicker, year, monthOfYear, dayOfMonth) -> {
            myCalendar.set(year, monthOfYear, dayOfMonth);
            purchaseDateNew.setText(dateFormat.format(myCalendar.getTime()));
        };
        purchaseDateNew.setOnClickListener(v1 -> {
            hideKeyboard(v1);
            new DatePickerDialog(this, date, myCalendar
                .get(Calendar.YEAR), myCalendar.get(Calendar.MONTH),
                myCalendar.get(Calendar.DAY_OF_MONTH)).show();
        });

        purchaseTitleNew.setOnKeyListener((view, i, keyEvent) -> {
            findViewById(R.id.purchasetitlenew).getBackground().setColorFilter(null);
            return false;
        });

        purchaseCreate.setOnClickListener(floatingButtonView -> {
            if (purchaseDateNew.getText().toString().equals("")) {
                purchaseDateNew.getBackground().setColorFilter(getResources().getColor(R.color.colorAccent), PorterDuff.Mode.SRC_IN);
                return;
            }

            hideKeyboard(floatingButtonView);

            Purchase newPurchase = new Purchase(purchaseDateNew.getText().toString(), purchaseTitleNew.getText().toString());
            DmServer.api.createUserPurchase(newPurchase).enqueue(new DmServer.Callback<Void>("createPurchase", this, true) {
                @Override
                public void onSuccessfulResponse(Response<Void> response) {
                    downloadPurchaseList();
                    AddIssue.this.toggleAddPurchaseButton(true);
                    AddIssue.this.showPurchases();
                    newPurchaseSection.setVisibility(View.GONE);
                }
            });

        });

        purchaseCreateCancel.setOnClickListener(floatingButtonView -> {
            hideKeyboard(floatingButtonView);

            newPurchaseSection.setVisibility(View.GONE);
            toggleAddPurchaseButton(true);
        });
    }

    private void hideKeyboard(View floatingButtonView) {
        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.hideSoftInputFromWindow(floatingButtonView.getWindowToken(), 0);
        }
    }

    private void showPurchases() {
        final RecyclerView rv = findViewById(R.id.purchase_list);
        rv.setAdapter(new PurchaseAdapter(this, purchases));
        rv.setLayoutManager(new LinearLayoutManager(this));
    }

    @Override
    public void onClick(View view) {
        switch(view.getId()) {
            case R.id.noCondition:
            case R.id.badCondition:
            case R.id.notSoGoodCondition:
            case R.id.goodCondition:
                ((TextView) findViewById(R.id.addissue_condition_text)).setText(view.getContentDescription().toString());
                break;
        }
    }
}

File: app/src/main/java/net/ducksmanager/whattheduck/AddIssue.kt
package net.ducksmanager.whattheduck

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.content.Context
import android.content.Intent
import android.graphics.PorterDuff
import android.os.Bundle
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.DatePicker
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import net.ducksmanager.api.DmServer
import net.ducksmanager.persistence.models.composite.InducksIssueWithUserIssueAndScore
import net.ducksmanager.persistence.models.composite.IssueListToUpdate
import net.ducksmanager.persistence.models.dm.Purchase
import net.ducksmanager.whattheduck.PurchaseAdapter.NoPurchase
import net.ducksmanager.whattheduck.databinding.AddissueBinding
import retrofit2.Response
import java.lang.ref.WeakReference
import java.text.SimpleDateFormat
import java.util.*

class AddIssue : AppCompatActivity(), View.OnClickListener {
    private lateinit var purchases: MutableList<Purchase>
    private lateinit var binding: AddissueBinding


    companion object {
        private val myCalendar = Calendar.getInstance()
        private val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = AddissueBinding.inflate(layoutInflater)
        setContentView(binding.root)
        downloadPurchaseList()
    }

    private fun downloadPurchaseList() {
        DmServer.api.userPurchases.enqueue(object : DmServer.Callback<List<Purchase>>("getpurchases", this, true) {
            override fun onSuccessfulResponse(response: Response<List<Purchase>>) {
                response.body()?.let {
                    WhatTheDuck.appDB.purchaseDao().deleteAll()
                    WhatTheDuck.appDB.purchaseDao().insertList(it)
                }
                setData()
            }
        })
    }

    private fun setData() {
        WhatTheDuck.appDB.purchaseDao().findAll().observe(this, Observer<List<Purchase>> { purchases: List<Purchase> ->
            this.purchases = arrayListOf(NoPurchase())
            this.purchases.addAll(purchases)

            show()
        })
    }

    private fun show() {
        binding.noCondition.setOnClickListener(this)
        binding.badCondition.setOnClickListener(this)
        binding.notSoGoodCondition.setOnClickListener(this)
        binding.goodCondition.setOnClickListener(this)

        showPurchases()
        binding.noCondition.performClick()

        PurchaseAdapter.selectedItem = purchases[0]

        binding.addissueOk.setOnClickListener {
            val dmCondition = when (binding.condition.checkedRadioButtonId) {
                R.id.badCondition -> InducksIssueWithUserIssueAndScore.BAD_CONDITION
                R.id.notSoGoodCondition -> InducksIssueWithUserIssueAndScore.NOTSOGOOD_CONDITION
                R.id.goodCondition -> InducksIssueWithUserIssueAndScore.GOOD_CONDITION
                else -> InducksIssueWithUserIssueAndScore.NO_CONDITION
            }
            val issueListToUpdate = IssueListToUpdate(
                WhatTheDuck.selectedPublication!!, listOf(WhatTheDuck.selectedIssue!!),
                dmCondition,
                if (PurchaseAdapter.selectedItem is NoPurchase) null else PurchaseAdapter.selectedItem!!.id
            )

            DmServer.api.createUserIssues(issueListToUpdate).enqueue(object : DmServer.Callback<Any>("addissue", this@AddIssue, true) {
                override fun onSuccessfulResponse(response: Response<Any>) {
                    finish()
                    WhatTheDuck.info(WeakReference(this@AddIssue), R.string.confirmation_message__issue_inserted, Toast.LENGTH_SHORT)
                    startActivity(Intent(this@AddIssue, Login::class.java))
                }
            })
        }

        binding.addissueCancel.setOnClickListener { finish() }

        binding.addpurchase.setOnClickListener {
            toggleAddPurchaseButton(false)
            showNewPurchase()
        }

        binding.addIssueTitle.text = getString(R.string.insert_issue__confirm, WhatTheDuck.selectedIssue)
    }

    private fun toggleAddPurchaseButton(toggle: Boolean) {
        binding.addpurchase.visibility =
            if (toggle)
                View.VISIBLE
            else
                View.GONE
    }

    private fun showNewPurchase() {
        val newPurchaseSection = binding.newpurchase
        newPurchaseSection.visibility = View.VISIBLE

        val purchaseDateNew = binding.purchasedatenew
        purchaseDateNew.requestFocus()
        purchaseDateNew.setText(dateFormat.format(Date()))
        purchaseDateNew.keyListener = null
        val date = OnDateSetListener { _: DatePicker?, year: Int, monthOfYear: Int, dayOfMonth: Int ->
            myCalendar[year, monthOfYear] = dayOfMonth
            purchaseDateNew.setText(dateFormat.format(myCalendar.time))
        }
        purchaseDateNew.setOnClickListener { v1: View ->
            hideKeyboard(v1)
            DatePickerDialog(this, date, myCalendar[Calendar.YEAR], myCalendar[Calendar.MONTH],
                myCalendar[Calendar.DAY_OF_MONTH]).show()
        }

        val purchaseTitleNew = binding.purchasetitlenew
        purchaseTitleNew.setOnKeyListener { _, _, _ ->
            binding.purchasetitlenew.background.colorFilter = null
            false
        }

        binding.createpurchase
            .setOnClickListener { floatingButtonView: View ->
                if (purchaseDateNew.text.toString() == "") {
                    purchaseDateNew.background.setColorFilter(ContextCompat.getColor(applicationContext, R.color.colorAccent), PorterDuff.Mode.SRC_IN)
                    return@setOnClickListener
                }

                hideKeyboard(floatingButtonView)

                val newPurchase = Purchase(purchaseDateNew.text.toString(), purchaseTitleNew.text.toString())
                DmServer.api.createUserPurchase(newPurchase).enqueue(object : DmServer.Callback<Void>("createPurchase", this, true) {
                    override fun onSuccessfulResponse(response: Response<Void>) {
                        downloadPurchaseList()
                        toggleAddPurchaseButton(true)
                        showPurchases()
                        newPurchaseSection.visibility = View.GONE
                    }
                })
            }

        binding.createpurchasecancel
            .setOnClickListener { floatingButtonView: View ->
                hideKeyboard(floatingButtonView)
                newPurchaseSection.visibility = View.GONE
                toggleAddPurchaseButton(true)
            }
    }

    private fun hideKeyboard(floatingButtonView: View) {
        (getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager)
            .hideSoftInputFromWindow(floatingButtonView.windowToken, 0)
    }

    private fun showPurchases() {
        val rv = binding.purchaseList
        rv.adapter = PurchaseAdapter(this, purchases)
        rv.layoutManager = LinearLayoutManager(this)
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.noCondition,
            R.id.badCondition,
            R.id.notSoGoodCondition,
            R.id.goodCondition ->
                binding.addissueConditionText.text = view.contentDescription.toString()
        }
    }
}